### PR TITLE
EOS-8251: check cas fom deadlock during hung_fom_notify

### PR DESCRIPTION
**Pull request by jadhavik1**
Exit the process when deadlock observed with foms so
failover/failback will restart the process.

Signed-off-by: Vikram Jadhav <vikram.jadhav@seagate.com>

### CODE REVIEW

**Comment by jadhavik1** [comment]
Changes are in testing, opened PR to get suggestions

File: cas/service.c [hunk 1]
```diff
@@ -1082,6 +1082,24 @@ static int op_sync_wait(struct m0_fom *fom)
...
+	struct cas_fom         *fom   = M0_AMB(fom, fom0, cf_fom);;
+	struct m0_long_lock    *lock  = m0_ctg_lock(fom->cf_ctg);
+	const struct m0_be_tx  *tx    = &fom0->fo_tx.tx_betx;
+	struct m0_be_tx_group  *gr    = tx->t_group;
+	enum m0_cas_opcode      opc   = m0_cas_opcode(fom0->fo_fop);
+	int                     phase = m0_fom_phase(fom0);
+
+	if (phase >= M0_FOPH_TYPE_SPECIFIC && 
+	    gr->tg_state == M0_BGS_FROZEN && !cas_is_ro(opc) &&
+	    ((phase == CAS_PREP && m0_be_tx_state(tx) == M0_BTS_GROUPING &&
```
Comment (sergey-shilov) [inline cas/service.c:1096]: Use M0_FOPH_TXN_WAIT instead of CAS_PREP.

File: cas/service.c [hunk 2]
```diff
@@ -1082,6 +1082,24 @@ static int op_sync_wait(struct m0_fom *fom)
...
+	      m0_long_is_write_locked(lock, fom0)) ||
+	     (m0_be_tx_state(tx) == M0_BTS_ACTIVE &&
```
Comment (sergey-shilov) [inline cas/service.c:1098]: Add check _phase == CAS_IDROP_LOCKED_

File: fop/fom.c [hunk 3]
```diff
@@ -335,11 +337,28 @@ M0_INTERNAL bool m0_fom_invariant(const struct m0_fom *fom)
 		_0C(fom->fo_service->rs_type == fom->fo_type->ft_rstype);
 }
 
+
+static void m0_fom_in_deadlock(const struct m0_fom *fom0)
+{
+	if (!M0_IN(fom0->fo_type->ft_id, (M0_CAS_GET_FOP_OPCODE,
```
Comment (sergey-shilov) [inline fop/fom.c:343]: M0_CAS_GET_FOP_OPCODE, M0_CAS_CUR_FOP_OPCODE and M0_CAS_REP_FOP_OPCODE are unnecessary here as they don't use tx. M0_CAS_GCW_FOP_OPCODE and M0_CAS_GCF_FOP_OPCODE also can be removed. PUT and DEL are enough here.

File: cas/service.c [hunk 4]
```diff
@@ -1082,6 +1082,24 @@ static int op_sync_wait(struct m0_fom *fom)
 	return M0_FSO_AGAIN;
 }
 
+M0_INTERNAL void m0_cas_fom_in_deadlock(const struct m0_fom *fom0)
```
Comment (sergey-shilov) [inline cas/service.c:1085]: Please return bool here, not exit(1).

File: fop/fom.c [hunk 5]
```diff
@@ -335,11 +337,28 @@ M0_INTERNAL bool m0_fom_invariant(const struct m0_fom *fom)
 		_0C(fom->fo_service->rs_type == fom->fo_type->ft_rstype);
 }
 
+
+static void m0_fom_in_deadlock(const struct m0_fom *fom0)
```
Comment (sergey-shilov) [inline fop/fom.c:341]: Please rename to m0_fom_check_deadlock().

File: fop/fom.c [hunk 6]
```diff
@@ -335,11 +337,28 @@ M0_INTERNAL bool m0_fom_invariant(const struct m0_fom *fom)
...
+	if (!M0_IN(fom0->fo_type->ft_id, (M0_CAS_GET_FOP_OPCODE,
+					  M0_CAS_PUT_FOP_OPCODE,
+					  M0_CAS_DEL_FOP_OPCODE,
+					  M0_CAS_CUR_FOP_OPCODE,
+					  M0_CAS_REP_FOP_OPCODE,
+					  M0_CAS_GCW_FOP_OPCODE,
+					  M0_CAS_GCF_FOP_OPCODE)))
+	    return;
+#ifndef __KERNEL__
+	m0_cas_fom_in_deadlock(fom0);
```
Comment (sergey-shilov) [inline fop/fom.c:352]: Please exit(1) if true returned.

**Review by sergey-shilov — CHANGES_REQUESTED** [review]

File: fop/fom.c [hunk 7]
```diff
@@ -335,11 +340,23 @@ M0_INTERNAL bool m0_fom_invariant(const struct m0_fom *fom)
...
 }
 
+static void m0_fom_check_deadlock(const struct m0_fom *fom)
+{
+	if (!M0_IN(fom->fo_type->ft_id, (M0_CAS_PUT_FOP_OPCODE,
+					 M0_CAS_DEL_FOP_OPCODE)))
+	    return;
+#ifndef __KERNEL__
+	if (m0_cas_fom_in_deadlock(fom))
+		exit(1);
```
Comment (huanghua78) [inline fop/fom.c:350]: Can you please print out some log messages to indicate that deadlock might be discovered and we have to exit now?
So, when this process exits, the HA will treat it as failure and will restart it again.

Comment (jadhavik1) [reply fop/fom.c:350]: yes will add log message, it is expected that HA will restart the process which got exited, so currently we are testing exit() behavior with HA on cluster, will update here once testing done.

**Comment by jadhavik1** [comment]
Following testcase done
1.Ran io 2 node cluster
2.Checked hctl status , all service started
3.Enabled error to exit from hun_fom_notify()
4.exit triggered for ioservice.
5.ioservice for that node gone offline
6.s3server service gone offline
7.later all started, its failback happened by HA
8.Same test conducted from another node

File: cas/service.c [hunk 8]
```diff
@@ -1082,6 +1082,31 @@ static int op_sync_wait(struct m0_fom *fom)
...
+M0_INTERNAL bool m0_cas_fom_in_deadlock(const struct m0_fom *fom0)
+{
+	struct cas_fom         *fom   = M0_AMB(fom, fom0, cf_fom);;
+	struct m0_long_lock    *lock  = m0_ctg_lock(fom->cf_ctg);
+	const struct m0_be_tx  *tx    = &fom0->fo_tx.tx_betx;
+	struct m0_be_tx_group  *gr    = tx->t_group;
+	enum m0_cas_opcode      opc   = m0_cas_opcode(fom0->fo_fop);
+	int                     phase = m0_fom_phase(fom0);
+
+	if (gr->tg_state == M0_BGS_FROZEN && !cas_is_ro(opc) &&
```
Comment (max-seagate) [inline cas/service.c:1099]: [* defect *] Access to `tg_state` is not protected by a lock. This means that the value may be obsolete even before the statement is fully  executed. Please ether fix or explain why this is acceptable here (if it is).

File: fop/fom.c (hunk 7 above)
Comment (max-seagate) [reply fop/fom.c:350]: Please use `M0_ASSERT()` or `M0_IMPOSSIBLE()` to clearly indicate that it's an abnormal termination of the process.

File: cas/service.c [hunk 9]
```diff
@@ -1082,6 +1082,31 @@ static int op_sync_wait(struct m0_fom *fom)
...
+	    ((phase == M0_FOPH_TXN_WAIT &&
+	      m0_be_tx_state(tx) == M0_BTS_GROUPING &&
+	      m0_long_is_write_locked(lock, fom0)) ||
```
Comment (max-seagate) [inline cas/service.c:1102]: [* doc *] Please explain in a comment why the use of this function is acceptable here.

In general the return value in such kind of functions is outdated even before the function returns.

**Review by max-seagate — CHANGES_REQUESTED** [review]

File: fop/fom.c (hunk 7 above)
Comment (jadhavik1) [reply fop/fom.c:350]: exit(1) was suggested by team,  not sure how HA behave with if is use M0_ASSERT

**Comment by jadhavik1** [comment]
@max-seagate  in this case of deadlock case engine has to be already locked, so removing lock taken

**Comment by jadhavik1** [comment]
Let me know if any other comments

**Comment by jadhavik1** [comment]
this PR can be considered if https://github.com/Seagate/cortx-motr/pull/57 not landed due to perf issue.

File: cas/service.c [hunk 10]
```diff
@@ -1082,6 +1082,38 @@ static int op_sync_wait(struct m0_fom *fom)
...
+M0_INTERNAL bool m0_cas_fom_in_deadlock(const struct m0_fom *fom0)
+{
+	struct cas_fom         *fom   = M0_AMB(fom, fom0, cf_fom);;
+	struct m0_long_lock    *lock  = m0_ctg_lock(fom->cf_ctg);
+	const struct m0_be_tx  *tx    = &fom0->fo_tx.tx_betx;
+	struct m0_be_tx_group  *gr    = tx->t_group;
+	enum m0_cas_opcode      opc   = m0_cas_opcode(fom0->fo_fop);
+	int                     phase = m0_fom_phase(fom0);
+
+	/**
```
Comment (huanghua78) [inline cas/service.c:1099]: Common comment starts with "/*".
No need to start with "/**".

File: fop/fom.c [hunk 11]
```diff
@@ -335,11 +337,25 @@ M0_INTERNAL bool m0_fom_invariant(const struct m0_fom *fom)
...
 }
 
+static void m0_fom_check_deadlock(const struct m0_fom *fom)
+{
+	if (!M0_IN(fom->fo_type->ft_id, (M0_CAS_PUT_FOP_OPCODE,
+					 M0_CAS_DEL_FOP_OPCODE)))
+	    return;
+#ifndef __KERNEL__
+	if (m0_cas_fom_in_deadlock(fom)) {
+		M0_LOG(M0_ERROR, "found deadlock with fom %p", fom);
```
Comment (huanghua78) [inline fop/fom.c:347]: s/M0_ERROR/M0_FATAL/

This is a fatal error. No chance to recover.

**Comment by jadhavik1** [comment]
PR57 Landed which has actual solution so closing this PR

**Pull request closed without merging**